doc/user-manual/language/core-language.lagda.rst

```diff
@@ -8,21 +8,182 @@
 Core language
 *************
 
-.. note::
-   This is a stub
+A program in Agda consists of a number of declarations written in an ``*.agda``
+file. A declaration introduces a new identifier and gives its type and
+definition. It is possible to declare:
+
+* :ref:`datatypes <data-types>`
+* :ref:`function definitions <function-definitions>`
+  (including :ref:`mixfix operators <mixfix-operators>` and
+  :ref:`abstract definitions <abstract-definitions>`)
+* local definitions :ref:`let <let-expressions>` and
+  :ref:`where <where-blocks>`
+* :ref:`record types <record-types>` (including
+  :ref:`coinductive records <copatterns-coinductive-records>`)
+* :ref:`modules <module-basics>`
+* :ref:`postulates <postulates>`
+* :ref:`variables <generalization-of-declared-variables>`
+* :ref:`pragmas <pragmas>`
+* :ref:`program options <command-line-options>`
+* :ref:`pattern-synonyms <pattern-synonyms>`, and
+* :ref:`precedence <precedence>` (fixity).
+
+Declarations have a signature part and a definition part. These can appear
+separately in the program. Names must be declared before they are used, but
+by separating the signature from the definition it is possible to define things
+in :ref:`mutual recursion <mutual-recursion>`.
+
+Grammar
+-------
+
+At its core, Agda is a dependently typed lambda calculus. The grammar of terms
+is as follows:
+
+.. code-block:: text
+
+  a ::= x                       -- variable
```

Comment:
This is incomplete.  Internally, we have everything that you show below with the `Term` type.

Reply:
I updated the grammar, but I am a bit lost when trying to translate the haskell term into BNF-style rules, can you help?

Reply:
We did some initial work on an Agda specification: https://agda.github.io/agda-spec/core-agda.pdf 
Maybe this could be a good source for the documentation you are writing here.